libr/bin/format/elf/elf.c

```diff
@@ -549,157 +530,178 @@ static void set_default_value_dynamic_info(ELFOBJ *bin) {
 	bin->dyn_info.dt_flags_1 = ELF_XWORD_MAX;
 	bin->dyn_info.dt_rpath = ELF_XWORD_MAX;
 	bin->dyn_info.dt_runpath = ELF_XWORD_MAX;
+	r_vector_init(&bin->dyn_info.dt_needed, sizeof(Elf_(Off)), NULL, NULL);
 }
 
-static int init_dynamic_section(ELFOBJ *bin) {
-	Elf_(Dyn) *dyn = NULL;
-	ut64 strtabaddr = 0;
-	char *strtab = NULL;
-	size_t strsize = 0;
-	int i, len, r;
-	ut8 sdyn[sizeof (Elf_(Dyn))] = { 0 };
-	ut64 dyn_size = 0, loaded_offset;
-	set_default_value_dynamic_info(bin);
+static size_t get_maximum_number_of_dynamic_entries(ut64 dyn_size) {
+	return dyn_size / sizeof (Elf_(Dyn));
+}
 
-	r_return_val_if_fail (bin, false);
-	if (!bin->phdr || !bin->ehdr.e_phnum) {
+static bool fill_dynamic_entry(ELFOBJ *bin, Elf_(Phdr) *dyn_phdr, ut64 dyn_size, size_t pos, Elf_(Dyn) *d) {
+	ut8 sdyn[sizeof (Elf_(Dyn))] = { 0 };
+	int j = 0;
+	int len = r_buf_read_at (bin->b, dyn_phdr->p_offset + pos * sizeof (Elf_(Dyn)), sdyn, sizeof (Elf_(Dyn)));
```

Comment:
the test `ELF: dynamic p_offset` is failing because you are using `dyn_phdr->p_offset` here instead of `loaded_offset` as computed in `init_dynamic_section`. You should use that value here.